### PR TITLE
interp: support pointer tagging

### PR DESCRIPTION
**Pull request by aykevl**
This is necessary to get https://github.com/tinygo-org/tinygo/pull/3691 working.

I have also refactored/fixed some of the related code to make the last commit a nice clean change.

### CODE REVIEW

File: interp/interpreter.go [hunk 1]
```diff
@@ -736,30 +736,25 @@ func (r *runner) run(fn *function, params []value, parentMem *memoryView, indent
 			if err == nil {
 				// The lhs is a pointer. This sometimes happens for particular
 				// pointer tricks.
-				switch inst.opcode {
-				case llvm.Add:
+				if inst.opcode == llvm.Add {
```
Comment (deadprogram) [inline interp/interpreter.go:739]: This looks like a good place for:

```go
switch {
    case inst.opcode == llvm.Add:
        // do stuff
    case inst.opcode == llvm.Xor && rhs.Uint() == 0:
        // other stuff
    default:
        // yet other stuff
}
```

Comment (deadprogram) [reply interp/interpreter.go:739]: I just noticed that it is not quite so easy to replace that code in this PR without changing how the commits are structured.

If I feel strongly I will submit a PR with that change.

Comment (aykevl) [reply interp/interpreter.go:739]: Oh, yes that's another option. Didn't think of that. It has the same effect though.

I don't really care about the specific form 🤷 I can update the PR if you like.

**Comment by deadprogram** [comment]
Thanks @aykevl for the improvements/fixes. This should unblock @dgryski 

Now merging.

**Pull request merged**